pilot/pkg/model/service.go

```diff
@@ -609,8 +609,16 @@ func ParseSubsetKey(s string) (direction TrafficDirection, subsetName string, ho
 
 // GetServiceAddressForProxy returns a Service's IP address specific to the cluster where the node resides
 func (s *Service) GetServiceAddressForProxy(node *Proxy) string {
-	if node.Metadata != nil && node.Metadata.ClusterID != "" && s.ClusterVIPs[node.Metadata.ClusterID] != "" {
-		return s.ClusterVIPs[node.Metadata.ClusterID]
+	clusterIP := func() string {
+		s.Mutex.RLock()
+		defer s.Mutex.RUnlock()
+		if node.Metadata != nil && node.Metadata.ClusterID != "" {
```

Comment:
Couple of points
- Should we move `if node.Metadata != nil && node.Metadata.ClusterID != ""` this outside so that only lock if cluserid is there ? It might reduce locking for some cases?
- If we lock here, are n't we locking on every call to `GetServiceAddressForProxy` - which is the very reason we moved that to push context so that we lock only once during initialization.

I could not think of better case to handle this at that time

Reply:
This is read lock, multi calls can happen concurrently.

Reply:
Shriram felt this is causing too much locks (read locks only) and he had some numbers in push timings. That is why we removed that lock https://github.com/istio/istio/pull/27027 here.

Are you trying to clean this up because it is not right to store it in push context?

Reply:
yep, and this info should not be duplicated. BTW, there is little chance to conflict if you look into it.